x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,216 @@
+    private static final float MAX_ORGANIC_SCORE = Float.intBitsToFloat((0xfe << 23)) - 1;
+
+    /**
+     * Creates a new PinnedQueryBuilder
+     */
+    public PinnedQueryBuilder(QueryBuilder organicQuery, String... ids) {
+      if (organicQuery == null) {
+          throw new IllegalArgumentException("[" + NAME + "] organicQuery cannot be null");
+      }
+      this.organicQuery = organicQuery;
+      if (ids == null) {
+          throw new IllegalArgumentException("[" + NAME + "] ids cannot be null");
+      }
+      if (ids.length > MAX_NUM_PINNED_HITS) {
+          throw new IllegalArgumentException("[" + NAME + "] Max of "+MAX_NUM_PINNED_HITS+" ids exceeded: "+
+                  ids.length+" provided.");          
+      }
+      for (String id : ids) {
+          if (id == null) {
+              throw new IllegalArgumentException("[" + NAME + "] id cannot be null");
+          }          
+          if(this.ids.add(id) == false) {
+              throw new IllegalArgumentException("[" + NAME + "] duplicate id found in list: "+id);              
+          }
+      }
+      
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PinnedQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        Collections.addAll(ids, in.readStringArray());
+        organicQuery = in.readNamedWriteable(QueryBuilder.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeStringArray(ids.toArray(new String[ids.size()]));
+        out.writeNamedWriteable(organicQuery);
+    }
+
+    /**
+     * @return the organic query set in the constructor
+     */
+    public QueryBuilder organicQuery() {
+        return this.organicQuery;
+    }
+
+    /**
+     * Returns the pinned ids for the query.
+     */
+    public Set<String> ids() {
+        return Collections.unmodifiableSet(this.ids);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (organicQuery != null) {
+            builder.field(ORGANIC_QUERY_FIELD.getPreferredName());
+            organicQuery.toXContent(builder, params);
+        }
+        builder.startArray(IDS_FIELD.getPreferredName());
+        for (String value : ids) {
+            builder.value(value);
+        }
+        builder.endArray();
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+    
+    
+    
+    private static final ConstructingObjectParser<PinnedQueryBuilder, XContentParser> PARSER = new ConstructingObjectParser<>(NAME,
+            a -> 
+                {
+                    QueryBuilder organicQuery = (QueryBuilder) a[0];
+                    @SuppressWarnings("unchecked")
+                    List<String> ids = (List<String>) a[1];
+                    return new PinnedQueryBuilder(organicQuery, ids.toArray(new String[0]));
```

Comment:
with recent versions of Java it is better practice to do `ids.toArray(String[]::new)` imo